src/NHibernate.Test/Linq/NullComparisonTests.cs

```diff
@@ -17,6 +18,320 @@ public class NullComparisonTests : LinqTestCase
+			Expect(q, Does.Contain("Input is null").IgnoreCase, BothDifferent, OutputSet, BothNull);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => o.Output != o.NullableAnotherEntityRequired.Input);
+			Expect(q, Does.Contain("Input is null").IgnoreCase, BothDifferent, OutputSet, BothNull);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => 3 != o.NullableOutput.Length);
+			Expect(q, Does.Contain("is null").IgnoreCase, InputSet, BothDifferent, BothNull, OutputSet);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => o.NullableOutput.Length != 3);
+			Expect(q, Does.Contain("is null").IgnoreCase, InputSet, BothDifferent, BothNull, OutputSet);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => 3 != o.Input.Length);
+			Expect(q, Does.Not.Contain("is null").IgnoreCase, InputSet, BothDifferent);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => o.Input.Length != 3);
+			Expect(q, Does.Not.Contain("is null").IgnoreCase, InputSet, BothDifferent);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => (o.NullableAnotherEntityRequiredId ?? 0) != (o.NullableAnotherEntityRequired.NullableAnotherEntityRequiredId ?? 0));
+			Expect(q, Does.Not.Contain("is null").IgnoreCase);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => (o.NullableAnotherEntityRequired.NullableAnotherEntityRequiredId ?? 0) != (o.NullableAnotherEntityRequiredId ?? 0));
+			Expect(q, Does.Not.Contain("is null").IgnoreCase);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => o.NullableAnotherEntityRequiredId.GetValueOrDefault() != o.NullableAnotherEntityRequired.NullableAnotherEntityRequiredId.GetValueOrDefault());
+			Expect(q, Does.Not.Contain("is null").IgnoreCase);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => o.NullableAnotherEntityRequired.NullableAnotherEntityRequiredId.GetValueOrDefault() != o.NullableAnotherEntityRequiredId.GetValueOrDefault());
+			Expect(q, Does.Not.Contain("is null").IgnoreCase);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => o.NullableAnotherEntityRequiredId.HasValue && o.NullableAnotherEntityRequired.NullableAnotherEntityRequiredId.HasValue && o.NullableAnotherEntityRequiredId.Value != o.NullableAnotherEntityRequired.NullableAnotherEntityRequiredId.Value);
+			Expect(q, Does.Not.Contain("is null").IgnoreCase);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => o.NullableAnotherEntityRequiredId.HasValue && o.NullableAnotherEntityRequiredId.Value != 0);
+			ExpectAll(q, Does.Not.Contain("is null").IgnoreCase);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => o.NullableAnotherEntityRequiredId.HasValue || o.NullableAnotherEntityRequiredId.Value != 0);
+			ExpectAll(q, Does.Contain("is null").IgnoreCase);
+
+			q = session.Query<AnotherEntityRequired>().Where(o => o.NullableOutput != null && o.NullableOutput != "test");
+			Expect(q, Does.Not.Contain("is null").IgnoreCase, BothDifferent, BothSame, OutputSet);
```

Comment:
Inverting tests fails here. Shouldn'it be supported too?
```
q = session.Query<AnotherEntityRequired>().Where(o => o.NullableOutput != "test" && o.NullableOutput != null);
Expect(q, Does.Not.Contain("is null").IgnoreCase, BothDifferent, BothSame, OutputSet);
```
I have seen your summary is just listing the case:

> The nullable property has a HasValue check before the != operator.

But having such a check after yields the same result set and it does not make much sense to have the `is null` check in SQL here.
From a Linq-To-Object stand-point, you can argue that Linq would first consider null as passing then the "and also" statement would ruled it out, so in some way this is translating to SQL what does Linq-To-Object. But I think we should align on Linq-To-Object at most about which result set is emitted. I do not think we should align on it also for how it gets it.

Has this choice been done because it would be overly complex to handle otherwise?

Reply:
> Inverting tests fails here. Shouldn'it be supported too?

It should be, but I guess I didn't considering them when I made this PR. The logic is now modified to support them.

> I have seen your summary is just listing the case
The nullable property has a HasValue check before the != operator.

Description updated.